### PR TITLE
Corrected cax attributes of ImageGrid axes

### PR DESCRIPTION
**Pull request by inclement**
Following the documentation, each axis of an ImageGrid should have a `cax` attribute pointing to the axis to be used for its colorbar. However, when the axes of a row or column share a colorbar, this attribute is not set correctly - colorbar axes are assigned as if every axis has its own colorbar, so most of the `cax` attributes are not visible axes, and the axes do not match the correct row or column.

This PR changes the behaviour to correctly associate axes of a row/column with the colorbar at the end of that row/column.

Example code demonsrating the problem is:

    from mpl_toolkits.axes_grid1 import ImageGrid
    import matplotlib.pyplot as plt
    import numpy as n

    X, Y = n.meshgrid(n.linspace(0, 6, 30), n.linspace(0, 6, 30))
    arr = n.sin(X) * n.cos(Y) + 1j*(n.sin(3*Y) * n.cos(Y/2.))

    fig = plt.figure(figsize=(6, 6))

    grid = ImageGrid(fig, 111,
                          nrows_ncols = (2, 2),
                          direction='row',
                          cbar_location='right',
                          cbar_mode='edge',
                     )
    ax1, ax2, ax3, ax4, = grid

    im1 = ax1.imshow(arr.real, cmap='spectral')
    im2 = ax2.imshow(arr.imag, cmap='hot')
    im3 = ax3.imshow(n.abs(arr), cmap='jet')
    im4 = ax4.imshow(n.arctan2(arr.imag, arr.real), cmap='hsv')

    ax2.cax.colorbar(im2)
    ax4.cax.colorbar(im4)

    fig.savefig('colorbar_failure.png')

Before this patch, the top-right colorbar will be empty while the bottom-right colorbar will show colours from one of the top-row plots. After this patch, the colorbars updated are those in the correct row/column to match the axis whose `cax` attribute was accessed.

This would be a breaking change for any users currently using the workaround of accessing the cax attributes or cbar_axes dircetly in order of their creation, which corresponds to the order in which they are visible. I'm not sure what the matplotlib policy is on such changes, let me know if it is inappropriate!

Edit: added results of the above script

before:
colorbar-current_behaviour

and after:
colorbar-new_behaviour

### CODE REVIEW

**Comment by inclement** [comment]
I merged the master branch into this one to fix a merge conflict that appeared, but now realise this was probably the wrong way to do it since it messes up the diff. Should I rebase a084e80 on the new master instead, or is there a better way? And is it okay to fix by force pushing to this branch, or should I make a new pr?

Sorry for the mistake, and thanks for the help!

**Comment by WeatherGod** [comment]
Rebase and force push. Feature branches are OK to force push because they
are understood to be in flux.

**Comment by tacaswell** [comment]
Could you also update the documentation on this function to explain the rules for sharing the color bars?

Am I understanding that the colors that end up in the color bar are selected based on the order in which they are called (so the calls that look like no-ops in the final figure are actually just getting replaced)?

re-milestoned as 1.5.x as this is seeming like a 'big' change to me now, but I am open to being (re)convinced it should be back-ported.

**Comment by inclement** [comment]
The core of the problem is that (by default), each axis of the ImageGrid receives a `cax` attribute that is supposed to refer to the axis on which its colorbar should be drawn. Calling `colorbar()` on this axis has the normal matplotlib behaviour in that repeat callings override the previous one. I can certainly add a note in the doc about this, but (if I understand right) it follows naturally from the normal colorbar behaviour. 

I only used the overriding in the test because it demonstrates the problem well - *without* this PR being applied, the colorbars will be displayed on the wrong axes or not displayed at all. Following this PR, the colorbar seen is the most recent one called from the same row/column as appropriate.

As for backporting, I think this is probably a legitimate bug based on the documentation and existing code (which seems to imply that the `cax` attribute of every axis should be the actual visible colorbar axis in the given row/column), but I can't say for sure and I don't know why this behaviour was omitted if so. It is a significant change in that it *will* break existing code that works around the problem. I'm not sure what the matplotlib policy is on such changes, but I hope this explanation makes my perspective clear on its nature.

File: lib/mpl_toolkits/tests/test_axes_grid.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
+
+from matplotlib.testing.decorators import image_comparison
+from mpl_toolkits.axes_grid1 import ImageGrid
+import numpy as np
+import matplotlib.pyplot as plt
+
+
+@image_comparison(baseline_images=['imagegrid_cbar_mode'],
+                  remove_text=True)
```
Comment (pelson) [inline lib/mpl_toolkits/tests/test_axes_grid.py]: I think we only need to test one format here. PNG is a good candidate.

**Comment by inclement** [comment]
@pelson Okay, I've set the test to png only

**Comment by tacaswell** [comment]
Can you re-base it to expunge the pdf/svg files from history?

**Comment by inclement** [comment]
I finally worked out how to drive an interactive rebase and I think I've successfully removed the pdf and svg from history.

**Comment by tacaswell** [comment]
Almost, it looks like 07a336e39f2b5c6b79cfece711e26d8ebc967524 removes an image added by  695f90e

Rebasing is tricky, but worth the learning curve (and wreckage along the way) ;).

**Comment by inclement** [comment]
Will it suffice to remove that image from the commit at which it was added but not change anything else, so the test will fail for those couple of commits (due to no reference image) but everything will be fixed by the later ones that modify the test and add the final image?

**Comment by tacaswell** [comment]
Yes.  We don't worry about tests passing on _every_ commit, only on the
merges to master/maintenance branches.

**Comment by inclement** [comment]
Okay, I think I've removed that image.

**Comment by tacaswell** [comment]
Looks good to me.  Would like feed back from @leejjoon before merging.

**Comment by tacaswell** [comment]
@inclement Thank you!

**Pull request merged**